AnnoDesigner/Helper/RoadSearchHelper.cs

```diff
@@ -0,0 +1,114 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Windows;
+using AnnoDesigner.Core.Models;
+
+namespace AnnoDesigner.Helper
+{
+    public static class RoadSearchHelper
+    {
+        private static void DoNothing(AnnoObject objectInRange) { }
+
+        public static Dictionary<double, Dictionary<double, AnnoObject>> PrepareGridDictionary(IEnumerable<AnnoObject> placedObjects)
+        {
+            var dictionary = new Dictionary<double, Dictionary<double, AnnoObject>>();
+
+            foreach (var placedObject in placedObjects)
+            {
+                var x = placedObject.Position.X;
+                var y = placedObject.Position.Y;
+                for (var i = 0; i < placedObject.Size.Width; i++)
+                {
+                    if (!dictionary.ContainsKey(x + i))
+                    {
+                        dictionary.Add(x + i, new Dictionary<double, AnnoObject>());
+                    }
+
+                    for (var j = 0; j < placedObject.Size.Height; j++)
+                    {
+                        dictionary[x + i][y + j] = placedObject;
+                    }
+                }
+            }
+
+            return dictionary;
+        }
+
+        public static HashSet<Point> BreadthFirstSearch(
+            IEnumerable<AnnoObject> placedObjects,
+            IEnumerable<AnnoObject> startObjects,
+            Func<AnnoObject, double> rangeGetter,
+            Action<AnnoObject> inRangeAction = null,
+            Dictionary<double, Dictionary<double, AnnoObject>> gridDictionary = null)
```

Comment:
I think these parameters should be reordered so that the first 3 parameters are `placedObjects`, `startObjects`, and `gridDictionary` as these are integral to the working of the function, whereas the `rangeGetter` and `inRangeAction` parameters should be optional.

Following on from the above - I would make the `rangeGetter` optional, as we are working directly with `AnnoObjects`, there's no need to have the lamda as a mandatory parameter.

Finally, I would probably make the `gridDictionary` a mandatory parameter, as it is a very expensive function to run, and it will force the programmer to think about what they are doing, rather than heedlessly calling `BreadthFirstSearch()`, not realising that `PrepareGridDictionary()` is being called every time.

Reply:
I disagree, gridDictionary can be ommited in cases you want it to construct it from placedObjects instead.
The rangeGetter doesn't always use AnnoObject's InfluenceRange property. I already call this function with influence range 1 below what it should be do correctly render the influence polygon. Also in the future this could be used to modify the range of public building based on palace level/other things that affect influence range.

I don't know about the performance impact of reconstructing the gridDictionary. It only iterates through a list and creates some instances of Dictionary. That should have minimum impact on performance. I'll do some performance testing and report.